### PR TITLE
Added data manager Record manipulation

### PR DESCRIPTION
**Pull request by gervandiepen**
This is the third part of the issue #1143 changes.
It contains the changes in DataManInfo and TableCopy to better manipulate the data manager Record (which is needed for the TaQL changes).

### CODE REVIEW

File: tables/DataMan/DataManInfo.cc [hunk 1]
```diff
@@ -207,6 +208,206 @@ Record DataManInfo::adjustStMan (const Record& dminfo, const String& dmType,
...
 }
 
+void DataManInfo::mergeInfo (Record& dminfo1, const Record& dminfo2)
+{
+  // See for each new data manager what to do.
+  for (uInt i2=0; i2<dminfo2.nfields(); ++i2) {
+    Record dm2 = dminfo2.subRecord(i2);
+    String type2;
+    String name2;
+    Vector<String> cols2;
```
Comment (mnijhuis-tos) [inline tables/DataMan/DataManInfo.cc:218]: Unused variable?

File: tables/DataMan/DataManInfo.cc [hunk 2]
```diff
@@ -207,6 +208,206 @@ Record DataManInfo::adjustStMan (const Record& dminfo, const String& dmType,
...
+          std::vector<String> colsnew;
+          // Keep column if not equal to one in new dm.
+          for (auto col2 : cols2) {
+            Bool fnd = False;
+            for (auto col : cols) {
+              if (col2 == col) {
+                fnd = True;
+                break;
+              }
+              if (!fnd) {
```
Comment (mnijhuis-tos) [inline tables/DataMan/DataManInfo.cc:287]: I don't understand the logic, but it looks like you should move this `if` out of the inner for loop.

File: tables/DataMan/test/tDataManInfo.cc [hunk 3]
```diff
@@ -0,0 +1,228 @@
+//# tDataManInfo.cc: Test program for class DataManInfo
+//# Copyright (C) 2021
```
Comment (mnijhuis-tos) [inline tables/DataMan/test/tDataManInfo.cc:2]: 2022? ;-)

File: tables/DataMan/test/tDataManInfo.cc [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+#include <casacore/tables/Tables/TableDesc.h>
+#include <casacore/tables/Tables/SetupNewTab.h>
+#include <casacore/tables/Tables/ScaColDesc.h>
+#include <casacore/tables/Tables/ArrColDesc.h>
+#include <casacore/casa/Containers/Record.h>
+#include <casacore/casa/Utilities/Assert.h>
+#include <casacore/casa/iostream.h>
+
+#include <casacore/casa/namespace.h>
+using namespace std;
```
Comment (mnijhuis-tos) [inline tables/DataMan/test/tDataManInfo.cc:38]: Google style forbids `using namespace xxx`, for good reasons, but I don't know if that also applies to Casacore.

File: tables/DataMan/test/tDataManInfo.cc [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+    cout << dminfo << remCols1 << endl;
+    Vector<String> remCols2 =
+      DataManInfo::removeDminfoColumns (dminfo, Vector<String>(1, "col1"));
+    cout << dminfo << remCols2 << endl;
+    DataManInfo::setTiledStMan (dminfo, Vector<String>(1, "col3"),
+                                "TiledShapeStMan", "TSMData", IPosition(3,3,4,5));
+    cout << dminfo << endl;
+}
+
+void mergeTest()
```
Comment (mnijhuis-tos) [inline tables/DataMan/test/tDataManInfo.cc:68]: The two scopes suggest this test is actually two independent tests: `mergeTestZeroFields()` and `mergeTestOneField`. The same argument holds for the finalizeTest, below.

**Review by mnijhuis-tos — CHANGES_REQUESTED** [review]

File: tables/DataMan/DataManInfo.cc (hunk 1 above)
Comment (gervandiepen) [reply tables/DataMan/DataManInfo.cc:218]: Removed

File: tables/DataMan/test/tDataManInfo.cc (hunk 5 above)
Comment (gervandiepen) [reply tables/DataMan/test/tDataManInfo.cc:68]: Split them

File: tables/DataMan/test/tDataManInfo.cc [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+  AlwaysAssertExit (dm1.asString("TYPE") =="abc");
+  AlwaysAssertExit (dm1.asString("NAME") == "col1");
+  AlwaysAssertExit (dm1.asArrayString("COLUMNS").size() == 1);
+  AlwaysAssertExit (dm1.asArrayString("COLUMNS").data()[0] == "col1");
+}
+
+void largeTest()
+{
+  // Create a table description with various data manager types and names.
+  TableDesc desc;
```
Comment (mnijhuis-tos) [inline tables/DataMan/test/tDataManInfo.cc:127]: `desc` is only defined but not used in this test...

**Review by mnijhuis-tos — CHANGES_REQUESTED** [review]

File: tables/DataMan/test/tDataManInfo.cc (hunk 3 above)
Comment (gervandiepen) [reply tables/DataMan/test/tDataManInfo.cc:2]: No, it has been created in 2021. I could add ,2022 but that is not necessary.

File: tables/DataMan/test/tDataManInfo.cc (hunk 4 above)
Comment (gervandiepen) [reply tables/DataMan/test/tDataManInfo.cc]: We'll stick to casacore guidelines

File: tables/DataMan/DataManInfo.cc [hunk 7]
```diff
@@ -207,6 +208,184 @@ Record DataManInfo::adjustStMan (const Record& dminfo, const String& dmType,
...
 
+void DataManInfo::mergeInfo (Record& dminfo1, const Record& dminfo2)
+{
+  // See for each new data manager what to do.
+  for (uInt i2=0; i2<dminfo2.nfields(); ++i2) {
+    Record dm2 = dminfo2.subRecord(i2);
+    String type2 (dm2.isDefined("TYPE")  ?  dm2.asString("TYPE") : String());
+    String name2 (dm2.isDefined("NAME")  ?  dm2.asString("NAME") : String());
+    // Add the data manager to the first, but overwrite if already there.
+    uInt inx = dminfo1.nfields();
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/DataMan/DataManInfo.cc:219]: Changed

File: tables/DataMan/DataManInfo.h [hunk 8]
```diff
@@ -119,6 +143,11 @@ class DataManInfo
 
   // Show the Table IO statistics.
   static void showDataManStats (const Table&, ostream&);
+
+private:
+  // Merge the columns mentioned in dm into dminfo, where dm is the possibly
+  // changed inxdm-th dm record of dminfo. Those columns will be merged into dm.
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/DataMan/DataManInfo.h:149]: Clarified it in TableUtil.h

File: tables/DataMan/DataManInfo.cc [hunk 9]
```diff
@@ -207,6 +208,184 @@ Record DataManInfo::adjustStMan (const Record& dminfo, const String& dmType,
...
+  for (uInt i=0; i<dminfo.nfields(); ++i) {
+    Record& subinfo = dminfo.rwSubRecord(i);
+    if (subinfo.isDefined("NAME")  &&  subinfo.isDefined("COLUMNS")) {
+      Vector<String> cols(subinfo.asArrayString("COLUMNS"));
+      if (! cols.empty()) {
+        String name = subinfo.asString(i);
+        for (uInt j=0; j<dmtab.nfields(); ++j) {
+          const Record& subtab = dmtab.subRecord(j);
+          if (subtab.isDefined("NAME")  &&  subtab.asString("NAME") == name) {
+            // Add column name to DM name to make it unique.
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/DataMan/DataManInfo.cc:380]: Good point. I've added a function to make the names unique.

File: tables/DataMan/test/tDataManInfo.cc (hunk 6 above)
Comment (gervandiepen) [reply tables/DataMan/test/tDataManInfo.cc:127]: Forgot to do the finalizeMerge, which I've added now.

File: tables/DataMan/test/tDataManInfo.cc [hunk 10]
```diff
@@ -211,6 +211,28 @@ void largeTest()
     AlwaysAssertExit (dminfoRes.subRecord(1).subRecord("SPEC").nfields() == 1);
     AlwaysAssertExit (dminfoRes.subRecord(1).subRecord("SPEC").asInt("BSZ") == 10);
   }
+  Record dminfoFinal = DataManInfo::finalizeMerge (desc, dminfoRes);
+  cout << dminfoFinal;
+}
+
+void testUniqueName()
```
Comment (mnijhuis-tos) [inline tables/DataMan/test/tDataManInfo.cc:218]: Please add a "FOO", "FOO", "FOO", "FOO" test, and check that the result is "FOO", "FOO_0", "FOO_1", "FOO_2".

File: tables/DataMan/test/tDataManInfo.cc [hunk 11]
```diff
@@ -219,20 +219,25 @@ void testUniqueName()
...
   DataManInfo::makeUniqueNames (dminfo);
-  AlwaysAssertExit (dminfo.subRecord(0).asString("NAME") == "EMPTY_1");
+  AlwaysAssertExit (dminfo.subRecord(0).asString("NAME") == "DM_1");
   AlwaysAssertExit (dminfo.subRecord(1).asString("NAME") == "nma");
-  AlwaysAssertExit (dminfo.subRecord(2).asString("NAME") == "nma_1");
-  AlwaysAssertExit (dminfo.subRecord(3).asString("NAME") == "EMPTY");
+  AlwaysAssertExit (dminfo.subRecord(2).asString("NAME") == "nma_2");
+  AlwaysAssertExit (dminfo.subRecord(3).asString("NAME") == "nma_1");
+  AlwaysAssertExit (dminfo.subRecord(4).asString("NAME") == "DM");
+  AlwaysAssertExit (dminfo.subRecord(5).asString("NAME") == "col1");
```
Comment (mnijhuis-tos) [inline tables/DataMan/test/tDataManInfo.cc:240]: Is this the desired behavior (DM_1 before DM)? Or are you planning further changes?

File: tables/DataMan/DataManInfo.cc [hunk 12]
```diff
@@ -336,35 +338,79 @@ Record DataManInfo::finalizeMerge (const TableDesc& desc, const Record& dminfo)
...
+  uInt suffix = 0;
+  Bool unique = False;
+  while (!unique) {
+    unique = True;
+    for (uInt i=0; i<dminfo.nfields(); ++i) {
+      if (Int(i) != excludeField) {
+        const Record& dm = dminfo.subRecord(i);
+        if (dm.isDefined("NAME")) {
+          String nm = dm.asString("NAME");
+          if (newName == nm) {
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/DataMan/DataManInfo.cc:408]: changed

File: tables/DataMan/DataManInfo.cc [hunk 13]
```diff
@@ -336,35 +338,79 @@ Record DataManInfo::finalizeMerge (const TableDesc& desc, const Record& dminfo)
...
     newdm.defineRecord (newdm.size(), dm);
   }
   return newdm;
 }
 
+void DataManInfo::makeUniqueNames (Record& dminfo)
+{
+  // Ensure that data manager names are unique by adding a suffix if needed.
+  // Empty names are initially set to 'EMPTY'.
+  // Start at the back, so the oldest entries keep their name.
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/DataMan/DataManInfo.cc:372]: Using a std::set to solve the problem

File: tables/DataMan/test/tDataManInfo.cc (hunk 10 above)
Comment (gervandiepen) [reply tables/DataMan/test/tDataManInfo.cc:218]: Done

File: tables/DataMan/test/tDataManInfo.cc (hunk 11 above)
Comment (gervandiepen) [reply tables/DataMan/test/tDataManInfo.cc:240]: Now in natural order

File: tables/DataMan/DataManInfo.cc [hunk 14]
```diff
@@ -383,11 +382,24 @@ void DataManInfo::makeUniqueNames (Record& dminfo)
...
+  }
+  // Now make the names unique if needed.
+  // The first instance is kept as is, others get suffix _1, _2, etc.
+  std::set<String> firstNames;
+  for (uInt i=0; i<dminfo.nfields(); ++i) {
+    Record& dm = dminfo.rwSubRecord(i);
+    String name = dm.asString("NAME");
+    if (firstNames.find(name) == firstNames.end()) {
+      // The first instance of this name is kept as is.
+      firstNames.insert (name);
```
Comment (mnijhuis-tos) [inline tables/DataMan/DataManInfo.cc:396]: Optional:
`std::set::insert` already tells if the name already existed, so the extra `find()` is not needed.
You can use:
```
auto inserted = firstNames.insert(name);
if (!inserted.second) {
  // The name already exists. Make it unique by adding a suffix.
  name = uniqueName (dminfo, name, i);
  dm.define ("NAME", name);
}
```

File: tables/DataMan/DataManInfo.cc [hunk 15]
```diff
@@ -383,11 +382,24 @@ void DataManInfo::makeUniqueNames (Record& dminfo)
...
+    } else {
+      // Make the name unique by adding the suffix.
+      String newName = uniqueName (dminfo, name, i);
+      if (newName != name) {
```
Comment (mnijhuis-tos) [inline tables/DataMan/DataManInfo.cc:400]: `(newName != name)` should always be true here, shouldn't it? I therefore removed it in my suggestion in my other comment.

**Review by mnijhuis-tos — CHANGES_REQUESTED** [review]

File: tables/DataMan/DataManInfo.cc (hunk 15 above)
Comment (gervandiepen) [reply tables/DataMan/DataManInfo.cc:400]: Correct; changed

**Review by mnijhuis-tos — APPROVED** [review]

**Pull request merged**